sopel/modules/safety.py

```diff
@@ -216,13 +218,31 @@ def toggle_safety(bot, trigger):
 # Code above also calls this if there are too many cache entries
 @sopel.module.interval(24 * 60 * 60)
 def _clean_cache(bot):
-    """Cleans up old entries in URL cache"""
-    # TODO: probably should use locks here, to make sure stuff doesn't explode
-    oldest_key_age = 0
-    oldest_key = ''
-    for key, data in sopel.tools.iteritems(bot.memory['safety_cache']):
-        if data['age'] > oldest_key_age:
-            oldest_key_age = data['age']
-            oldest_key = key
-    if oldest_key in bot.memory['safety_cache']:
-        del bot.memory['safety_cache'][oldest_key]
+    """Cleans up old entries in URL safety cache."""
+    if bot.memory['safety_cache_lock'].acquire(False):
+        LOGGER.info('Starting safety cache cleanup...')
+        # clean up by age first
+        cutoff = time.time() - (7 * 24 * 60 * 60)  # 7 days ago
+        old_keys = []
+        for key, data in sopel.tools.iteritems(bot.memory['safety_cache']):
+            if data['fetched'] <= cutoff:
+                old_keys.append(key)
+        for key in old_keys:
+            bot.memory['safety_cache'].pop(key, None)
+
+        # clean up more values if the cache is still too big
+        overage = bot.memory['safety_cache'] - cache_limit
+        if overage > 0:
+            extra_keys = sorted(
+                (data.fetched, key)
+                for (key, data)
+                in bot.memory['safety_cache'].items())[:overage]
+            for (_, key) in extra_keys:
+                bot.memory['safety_cache'].pop(key, None)
+
+        LOGGER.info('Safety cache cleanup finished.')
+        bot.memory['safety_cache_lock'].release()
```

Comment:
What if any of the above code raise an exception and `release` is never called? This might need a try/except block.

Also, if a job is running twice at the same time with a 24h interval, we have a bigger problem than having to wait for the lock to be available.

Reply:
It's also run directly if the cache gets too big. The interval is only one way it can get called.

Fair point about `release`. I probably should've gone with my initial impulse and used a `with`.

Reply:
Actually I remember now why I didn't use `with`… I need the return value from `acquire` in non-blocking mode to log that debug entry. Which isn't _strictly_ necessary, but…

Reply:
I used a try/finally instead.